### ISSUE TITLE
Unknown key with X keycode... error on RHEL7 accessed via OS X Sierra

### ISSUE THREAD

**Issue by shadowguy** [FreeRDP/FreeRDP#4215]
Trying to use xfreerdp on my organization's RHEL7 server, using a Macbook Pro with OS X Sierra (10.12.6).  I cannot type anything as I get the "Unknown key with

**Comment by akallabeth** [comment]
@shadowguy Care to enlighten us with some more details? (FreeRDP version, command line, ...)?
https://github.com/FreeRDP/FreeRDP/wiki/BugReporting

**Comment by shadowguy** [comment]
Sorry about that @akallabeth, it's a trunk build that was built on Oct 27, reported version is 2.0.0-dev (510bf82).  The command line used is 'xfreerdp -sec-nla /d: /size:90% /drive:PLC,/reg/g/pcds/plc-common /v:plc-prg-03' where plc-prg-03 is a 64-bit Windows 7 Enterprise machine.  We turn off NLA because these are machines which are infrequently accessed and the passwords often are expired.

As mentioned, I encountered this using a Macbook to ssh to my organization's RHEL7 machine from which I then use xfreerdp to access the Windows node.  When I get home, I can check if it happens with a simplified command line.

**Comment by akallabeth** [comment]
Ok, so you're using X11 forwarding to display xfreerdp on you macbook but it runs on your RHEL7? Looks like the (already tricky) keyboard to scancode conversion is getting screwed by that.

**Comment by akallabeth** [comment]
After giving this some thought, we'd accept pull requests fixing this issue, but will not actively try to resolve it.

**Comment by shadowguy** [comment]
Yes, that is correct.  Unfortunately there is no other way to access these machines from outside my organization, they are kept on a subnet isolated from the internet.

**Comment by shadowguy** [comment]
@akallabeth Another data point, I just upgraded to High Sierra 10.13.1 and also installed the Xcode developer tools, I still run into this issue.

**Comment by shadowguy** [comment]
Another data point and possible workaround: when I use the 'virtual desktop' rather than 'floating window' option in NoMachine to connect to my organization the problem goes away!

**Comment by bmiklautz** [comment]
I hit the same problem but haven't found a good solution yet. The thing is that when compiled on Linux the mac keyboard support isn't compiled in and therefore the forwarded xfreerdp doesn't know the keys.

**Comment by teddyrendahl** [comment]
Same issue 👍

**Comment by MarkLTZ** [comment]
I have the same issue on Centos8 running xfreedrp via VNC Server.

xfreerdp /kbd-lang:0x0410 /compression /cert-ignore /u:xxxxx /p:xxxxx /v:xx.xx.xx.xx

Italian keyboard is not loaded correctly and I'm not able to use char like:
@òàùè+ì'?^é*ç€
and may be more chars

Running xfreerdp via gnome-session works, but I need to encapsulate it inside a vnc session where only xfreerdp is running.

Could someone help me to solve it, please?
Thanks

**Comment by DanFulger** [comment]
I think you could fix this in the VNC server by using RemapKeys ("0x0028->0x0039,0x0029->0x0030"). But I have not tried this.
What I have tried, and works, is this quick and dirty patch:
[FreeRDP_source.txt](https://github.com/FreeRDP/FreeRDP/files/4338383/FreeRDP_source.txt)

A better patch would try to search the current keyboard layout for the key that has the parenthesis symbol on it. But this is really a bug in VNC.

**Comment by MarkLTZ** [comment]
Which key fix your patch?

**Comment by DanFulger** [comment]
Sorry, I misread your comment. My fix is about parenthesis.

**Comment by MarkLTZ** [comment]
No problem, I missed your solution too unti yesterday.

Which parenthesis ( ) or [ ] ?

**Comment by DanFulger** [comment]
( and ).

**Comment by MarkLTZ** [comment]
ohh ok but these works before. 

My issue is on @ ò à ù è + ì ' ? ^ é * ç € [ ]
